### PR TITLE
Canonicalize Eltwise binary ops for broadcast

### LINKED ISSUE #815: TTNN throws an exception if operands require reordering for broadcast
In a case where the operand requires broadcast, TTNN assumes the positioning of operands to be canonicalized and throws the following exception otherwise

RuntimeError: TT_THROW @ tt-metal/src/tt-metal/ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.cpp:46: tt::exception

`TT_THROW("ttnn::operations::binary::BinaryDeviceOperation: unsupported broadcast");`

Apply canonicalization on the operands to allow broadcasting. 

If canonicalization is not possible (i.e. sub(a,b) != sub(b,a) we could always broadcast to nop (i.e. mul(a, 1)) and then do the eltwise.

### PR DESCRIPTION
**Pull request by uazizTT**
Canonicalize Eltwise binary ops by switching operands in case the `_rhs_` operand requires broadcast. 
Make SubtractOp commutative by converting it to combination of `NegOp` and `AddOp`: 
`SubtractOp(lhs, rhs) -> AddOp(lhs, NegOp(rhs)`

Fixes #815

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

**Comment by svuckovicTT** [comment]
I think this is a perfect candidate for https://mlir.llvm.org/docs/Canonicalization/

**Review by jnie-TT — APPROVED** [review]

File: runtime/lib/ttnn/operations/eltwise/binary.cpp
Comment (nsmithtt) [inline runtime/lib/ttnn/operations/eltwise/binary.cpp:20]: See comment above.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp
Comment (mrakitaTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:683]: Yeah, this is another example of what we've talked about. @uazizTT to give you more context, one of the topics of discussion in F2F meetings this week was how to handle metal issues that we hit on our side. We've agreed that we need to make a mindset shift where we first talk with metal team and see if they can fix it on their end, before making workarounds on our side. So this is the desired pattern of handling issues like this one:
1. Open issue in metal repo and ask for estimate on how long it will take to fix.
2. If they agree to fix it in a reasonable time, then we just wait. If not, then ask how feasible it is to fix, i.e. how much work they think it would take so we can see if we want to help with a fix.
3. If we are okay with fixing it ourselves and making a contribution then we do that. If not, then we figure out what is the best workaround to make on our end.
4. We open issue on our end to revert the workaround when the fix on metal side lands.

@uazizTT Can you please open issues on metal repo (from step 1) and on our repo (from step 4)? If you don't get any response from metal team and need to escalate, you can tag @tt-mpantic to assist.

Comment (mrakitaTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:683]: Btw @uazizTT, another thing that we've discussed in F2F meetings was how to improve our PR process. From now on it will be required to always get a sign off from at least one code owner of each code component that you've changed. This will be automated by github rules and we also have to update the codeowners file, but until then, could you please ping the codeowners who didn't approve before you merge just to check if they are okay with that, especially if there are some unresolved comments that they've left. This week many of us were too busy with F2F meetings to review PRs so I understand that you felt blocked, don't worry about it nobody will take it amiss, also this is a fresh thing that we've just discussed and old rules were more relaxed 🙂 The main point is that we are not in a rush to make changes asap and we should invest more time to make sure that we are making the best changes we can, so even if you have to wait a few days to iterate and get approval - in most cases it will be fine.

**Comment by uazizTT** [comment]
Filed the following issue to remove this workaround when TT-Metal adds support.

https://github.com/tenstorrent/tt-mlir/issues/868

File: runtime/lib/ttnn/operations/eltwise/binary.cpp
Comment (sdjordjevicTT) [inline runtime/lib/ttnn/operations/eltwise/binary.cpp:19]: Can we please swap introducing the workarounds in the runtime? This is a great example of how we could utilize canonicalization in MLIR...

**Pull request merged**
